### PR TITLE
BREAKING -- Support multiple git SHA targets

### PR DESCRIPTION
**Pull request by jjcomer**
Changed the API routes to support targetting particular git shas for configuration.
Added caching of know SHAs within an LRU

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -226,30 +277,30 @@ enum Config {
...
+            Config::Environment(e) => Some((*e).clone()),
             _ => None,
         }
     }
 
-    fn as_environment_type(self) -> Option<EnvironmentType> {
+    fn as_environment_type(&self) -> Option<EnvironmentType> {
         match self {
-            Config::EnvironmentType(e) => Some(e),
+            Config::EnvironmentType(e) => Some((*e).clone()),
```
Comment (jonathanmorley) [inline src/config.rs]: Why do we need to change these?

File: src/main.rs [hunk 2]
```diff
@@ -1,5 +1,3 @@
-#![warn(unused)]
```
Comment (jonathanmorley) [inline src/main.rs:1]: Why did these get removed?

File: src/transform/helper_comma_delimited_list.rs [hunk 3]
```diff
@@ -38,9 +37,9 @@ impl HelperDef for CommaDelimitedListHelper {
                         }
 
                         if let Some(block_param) = h.block_param() {
-                            let mut map = BTreeMap::new();
-                            map.insert(block_param.to_string(), to_json(item));
-                            local_rc.push_block_context(&map)?;
+                            let mut block_params = BlockParams::new();
+                            block_params.add_value(&block_param.to_string(), to_json(item))?;
```
Comment (jonathanmorley) [inline src/transform/helper_comma_delimited_list.rs:41]: What is the difference here?

File: src/config.rs [hunk 4]
```diff
@@ -139,21 +146,65 @@ impl ConfigDir {
...
+                            "Error: {} \n Unable to find the git repo: {}",
+                            e,
+                            directory.to_str().unwrap()
+                        );
+                        return None;
+                    }
+                };
+                match git::reset(
+                    &git_repo,
+                    "origin/master",
```
Comment (jonathanmorley) [inline src/config.rs]: This should probably be the default branch rather than hardcoded to master

File: src/git.rs [hunk 5]
```diff
@@ -38,3 +38,99 @@ pub fn clone(
...
+
+pub fn build_repo(path: &str) -> Result<Repository, Error> {
+    Repository::open(path).map_err(|e| e.into())
+}
+
+pub fn get_head_sha(repo: &Repository) -> Result<String, Error> {
+    let head = repo.head()?;
+    if let Some(target) = head.target() {
+        let sha = target.to_string();
+        Ok(sha[..7].to_string())
```
Comment (jonathanmorley) [inline src/git.rs:213]: Do we need to truncate?

File: src/config.rs (hunk 1 above)
Comment (jjcomer) [reply src/config.rs]: Was recommended by clippy

File: src/main.rs (hunk 2 above)
Comment (jjcomer) [reply src/main.rs:1]: They are on by default now. We get warnings that we are doubling down on overrides

File: src/transform/helper_comma_delimited_list.rs (hunk 3 above)
Comment (jjcomer) [reply src/transform/helper_comma_delimited_list.rs:41]: The interface changed with the new handlebars lib.

File: src/config.rs (hunk 4 above)
Comment (jjcomer) [reply src/config.rs]: We don't leave it on master, all routes require a sha to be specified. I can see if the lib will let me derive the default branch.

File: src/git.rs (hunk 5 above)
Comment (jjcomer) [reply src/git.rs:213]: Don't have to but wanted to be consistent with what we would expect from a user on the api

File: src/config.rs (hunk 1 above)
Comment (jjcomer) [reply src/config.rs]: Right. Clippy says we should not be using a `&self` as a parameter. Changing our self to a ref prevented us from being able to return a non-reference.

Comment (jonathanmorley) [reply src/config.rs]: I'll take a look, but perhaps it's as simple as a bad naming convention for the method

Comment (jonathanmorley) [reply src/config.rs]: https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention implies that we can rename these to into_foo instead

File: src/config.rs (hunk 4 above)
Comment (jonathanmorley) [reply src/config.rs]: Any update here?

**Pull request merged**